src/ra_log_segment.erl

```diff
@@ -258,10 +261,62 @@ read(State, Idx, Num) ->
 -spec read_cons(state(), ra_index(), Num :: non_neg_integer(),
                 fun((binary()) -> term()), Acc) ->
     Acc when Acc :: [{ra_index(), ra_term(), binary()}].
-read_cons(#state{cfg = #cfg{fd = _Fd, mode = read},
-                 index = Index} = State, Idx,
-          Num, Fun, Acc) ->
-    pread_cons(State, Idx, Idx + Num - 1, Index, Fun, Acc).
+read_cons(#state{cfg = #cfg{mode = read} = Cfg,
+                 cache = Cache,
+                 index = Index}, Idx, Num, Fun, Acc) ->
+    pread_cons(Cfg, Cache, Idx, Idx + Num - 1, Index, Fun, Acc).
+
+read_sparse(#state{index = Index,
+                   cfg = Cfg,
+                   cache = _Cache0}, Indexes, Fun, Acc) ->
+    Cache0 = prepare_cache(Cfg, Indexes, Index),
+    Entries = read_sparse0(Cfg, Indexes, Index, Cache0, Fun, Acc),
+    {undefined, length(Entries), Entries}.
+
+read_sparse0(_Cfg, [], _Index, _Cache, _Fun, Acc) ->
+    Acc;
+read_sparse0(Cfg, [NextIdx | Rem] = Indexes, Index, Cache0, Fun, Acc) ->
+    {Term, Offset, Length, _} = map_get(NextIdx, Index),
+    case cache_read(Cache0, Offset, Length) of
+        false ->
+            case prepare_cache(Cfg, Indexes, Index) of
+                undefined ->
+                    {ok, Data, _} = pread(Cfg, undefined, Offset, Length),
+                    read_sparse0(Cfg, Rem, Index, undefined, Fun,
+                                 [{NextIdx, Term, Fun(Data)} | Acc]);
+                Cache ->
+                    read_sparse0(Cfg, Indexes, Index, Cache, Fun, Acc)
+            end;
+        Data ->
+            read_sparse0(Cfg, Rem, Index, Cache0, Fun,
+                         [{NextIdx, Term, Fun(Data)} | Acc])
+    end.
+
+cache_read({CPos, CLen, Bin}, Pos, Length)
+  when Pos >= CPos andalso
+       Pos + Length =< (CPos + CLen) ->
+    %% read fits inside cache
+    binary:part(Bin, Pos - CPos, Length);
+cache_read(_, _, _) ->
+    false.
+
+prepare_cache(#cfg{} = _Cfg, [_], _SegIndex) ->
+    undefined;
+prepare_cache(#cfg{fd = Fd} = _Cfg, [First | Rem], SegIndex) ->
+    case consec_run(First, First, Rem) of
+        {I, I} ->
+            %% no run, no cache;
+            undefined;
+        {FstI, LastI} ->
+            {_, FstOffset, FstLength, _} = map_get(FstI, SegIndex),
+            {_, LastOffset, LastLength, _} = map_get(LastI, SegIndex),
+            %% The cache needs to be at least as large as the next entry
+            %% but no larger than ?READ_AHEAD_B
+            MaxCacheLen = FstLength + LastOffset + LastLength,
+            CacheLen = min(MaxCacheLen, max(FstLength, ?READ_AHEAD_B)),
```

Comment:
I don't understand this computation. If you are about to read `(Max)CacheLen` starting from `FstOffset`, doesn't `LastOffset` already include `FstOffset + FstLength` already?

I mean, if the maximum length to read should go from `FstOffset` to `LastOffset + LastLength`, then I would use:
```erlang
MaxCacheLen = LastOffset - FstOffset + LastLength
```
`LastOffset - FstOffset` corresponding to the area between the two offset and thus includes `FstOffset`.

The current computation of `MaxCacheLen` seems to be too large to me.

I do understand that you want `CacheLen` to be at least `FstLength` bytes to make sure that the entry is read entirely. But the logic doesn't correpond to the intention, at least to me, even though the result is probably correct in the end (in terms of the number of bytes read). Making sure that at least `FstLength` are read should be in the computation of `CacheLen`.

I'm not sure my argument is clear :-)